engine/src/core/error/expected.h

```diff
@@ -0,0 +1,108 @@
+//
+// Created by grzegorz on 18.09.19.
+//
+
+#pragma once
+
+#include <utility>
+#include "unexpected.h"
+#include "bald_assert.h"
+
+
+namespace Bald {
+    template<typename E, typename U>
+    class expected {
+        union {
+            E valid;
+            U invalid;
+        };
+        bool _isOk = true;
+    public:
+        constexpr expected() { new(&valid) E(); }
+
+        constexpr expected(const E& rhs) { new(&valid) E(rhs); }
+
+        constexpr expected(const unexpected<U>& rhs) :
+                _isOk(false) { new(&invalid) U(rhs.value); }
+
+        constexpr expected(E&& rhs) { new(&valid) E(std::forward<E>(rhs)); }
```

Comment:
Shouldn't std::move be used here?

Reply:
no difference but the move is more readable